tests/src/tests.rs

```diff
@@ -570,73 +597,216 @@ fn test_part(
+                    });
+                }
+            }
 
-    unexpected_outputs.sort_by_key(|&v| v.range.as_ref().map(|r| r.start));
-    missing_outputs.sort_by_key(|&v| v.range.as_ref().map(|r| r.start));
+            actual_diagnostics.insert(annotation);
+        }
 
-    // This prints all unexpected emits first, then all missing emits.
-    // Is this reasonable or subject to change?
-    if !(unexpected_outputs.is_empty() && missing_outputs.is_empty()) {
-        writeln!(output, "  Subtest {i} does not match expected errors.").unwrap();
-        ok = false;
+        // Basically symmetric_difference, but we need to know where an item is coming from.
+        let mut unexpected_outputs = actual_diagnostics
+            .difference(&diagnostic_annotations)
+            .collect::<Vec<_>>();
+        let mut missing_outputs = diagnostic_annotations
+            .difference(&actual_diagnostics)
+            .collect::<Vec<_>>();
+
+        unexpected_outputs.sort_by_key(|&v| v.range.as_ref().map(|r| r.start));
+        missing_outputs.sort_by_key(|&v| v.range.as_ref().map(|r| r.start));
+
+        // This prints all unexpected emits first, then all missing emits.
+        // Is this reasonable or subject to change?
+        if !(unexpected_outputs.is_empty() && missing_outputs.is_empty()) {
+            writeln!(output, "  Subtest {i} does not match expected errors.").unwrap();
+            ok = false;
+
+            for unexpected in unexpected_outputs {
+                write!(output, "    Not annotated | ").unwrap();
+                print_annotation(output, &source, line, unexpected)
+            }
 
-        for unexpected in unexpected_outputs {
-            write!(output, "    Not annotated | ").unwrap();
-            print_annotation(output, &source, line, unexpected)
+            for missing in missing_outputs {
+                write!(output, "    Not emitted   | ").unwrap();
+                print_annotation(output, &source, line, missing)
+            }
+        }
+    } else {
+        // warns and ignores diagnostics
+        if !diagnostic_annotations.is_empty() {
+            writeln!(
+                output,
+                "  Subtest {i} contains diagnostics but is in autocomplete mode."
+            )
+            .unwrap();
+            for a in diagnostic_annotations {
+                write!(output, "    Ignored | ").unwrap();
+                print_annotation(output, &source, line, &a);
+            }
         }
 
-        for missing in missing_outputs {
-            write!(output, "    Not emitted   | ").unwrap();
-            print_annotation(output, &source, line, missing)
+        for annotation in metadata.annotations.iter().filter(|a| {
+            matches!(
+                a.kind,
+                AnnotationKind::AutocompleteContains
+                    | AnnotationKind::AutocompleteExcludes
+            )
+        }) {
+            let cursor = if let Some(range) = &annotation.range {
```

Comment:
could we throw those warnings during metadata parsing? this function is already _really_ complicated.